File: src/logic.ts
import { Card, InsertEdge, RPS } from './types';

const WIN_MAP: Record<RPS, RPS | null> = {
  [RPS.ROCK]: RPS.SCISSORS,
  [RPS.SCISSORS]: RPS.PAPER,
  [RPS.PAPER]: RPS.ROCK,
  [RPS.BLANK]: null
};

const SCORE_WEIGHTS: Record<RPS, number> = {
  [RPS.ROCK]: 4,
  [RPS.SCISSORS]: 3,
  [RPS.PAPER]: 1,
  [RPS.BLANK]: 0
};

export interface LaneResult {
  newGrid: RPS[][];
  scoreDelta: number;
  penalty: number;
  laneScores: number[];
  replacedCells: { r: number; c: number }[];
  attachmentOffset: number;
  shiftedLanes: { index: number; type: 'row' | 'col'; direction: 1 | -1 }[];
}

function shiftLane(grid: RPS[][], index: number, type: 'row' | 'col', direction: 1 | -1): void {
  const size = grid.length;
  if (type === 'row') {
    const row = grid[index];
    if (direction === 1) {
      const last = row[size - 1];
      for (let col = size - 1; col > 0; col -= 1) {
        row[col] = row[col - 1];
      }
      row[0] = last;
    } else {
      const first = row[0];
      for (let col = 0; col < size - 1; col += 1) {
        row[col] = row[col + 1];
      }
      row[size - 1] = first;
    }
  } else {
    if (direction === 1) {
      const last = grid[size - 1][index];
      for (let row = size - 1; row > 0; row -= 1) {
        grid[row][index] = grid[row - 1][index];
      }
      grid[0][index] = last;
    } else {
      const first = grid[0][index];
      for (let row = 0; row < size - 1; row += 1) {
        grid[row][index] = grid[row + 1][index];
      }
      grid[size - 1][index] = first;
    }
  }
}

export function resolveAttachmentOffset(matrixSize: number, cardLength: number, pointerRatio: number): number {
  const minOffset = Math.min(0, matrixSize - cardLength);
  const maxOffset = Math.max(0, matrixSize - cardLength);

  if (minOffset === maxOffset) {
    return minOffset;
  }

  const placementCount = maxOffset - minOffset + 1;
  const clampedRatio = Math.min(Math.max(pointerRatio, 0), 0.999999);
  const placementIndex = Math.min(placementCount - 1, Math.floor(clampedRatio * placementCount));
  return minOffset + placementIndex;
}

export function executeLaneClash(
  currentGrid: RPS[][],
  edge: InsertEdge,
  card: Card,
  attachmentOffset: number
): LaneResult {
  const newGrid = currentGrid.map((row) => [...row]);
  const size = newGrid.length;
  let totalScore = 0;
  let penalty = 0;
  const laneScores: number[] = Array.from({ length: card.symbols.length }, () => 0);
  const replacedCells: { r: number; c: number }[] = [];
  const shiftedLanes: { index: number; type: 'row' | 'col'; direction: 1 | -1 }[] = [];

  for (let cardIndex = 0; cardIndex < card.symbols.length; cardIndex += 1) {
    const laneIndex = attachmentOffset + cardIndex;
    if (laneIndex < 0 || laneIndex >= size) {
      continue;
    }

    const attacker = card.symbols[cardIndex];
    let r = 0, c = 0, dr = 0, dc = 0;

    if (edge === 'TOP') { r = 0; c = laneIndex; dr = 1; dc = 0; }
    else if (edge === 'BOTTOM') { r = size - 1; c = laneIndex; dr = -1; dc = 0; }
    else if (edge === 'LEFT') { r = laneIndex; c = 0; dr = 0; dc = 1; }
    else if (edge === 'RIGHT') { r = laneIndex; c = size - 1; dr = 0; dc = -1; }

    const defender = newGrid[r][c];
    const attackerLoses = defender !== RPS.BLANK && (attacker === RPS.BLANK || WIN_MAP[defender] === attacker);

    if (attackerLoses) {
      const attackerVal = Number(SCORE_WEIGHTS[attacker]) || 0;
      penalty += attackerVal;

      if (edge === 'LEFT' || edge === 'RIGHT') {
        const shiftDir: 1 | -1 = edge === 'LEFT' ? -1 : 1;
        shiftLane(newGrid, laneIndex, 'row', shiftDir);
        shiftedLanes.push({ index: laneIndex, type: 'row', direction: shiftDir });
      } else {
        const shiftDir: 1 | -1 = edge === 'TOP' ? -1 : 1;
        shiftLane(newGrid, laneIndex, 'col', shiftDir);
        shiftedLanes.push({ index: laneIndex, type: 'col', direction: shiftDir });
      }
    } else if (attacker !== RPS.BLANK) {
      for (let step = 0; step < size; step += 1) {
        const currentDefender = newGrid[r][c];
        const attackerWins = (WIN_MAP[attacker] === currentDefender) || (currentDefender === RPS.BLANK);

        if (attackerWins) {
          const gain = Number(SCORE_WEIGHTS[currentDefender]) || 0;
          totalScore += gain;
          laneScores[cardIndex] += gain;
          newGrid[r][c] = attacker;
          replacedCells.push({ r, c });
          r += dr; c += dc;
          if (r < 0 || r > size - 1 || c < 0 || c > size - 1) break;
        } else {
          break;
        }
      }
    }
  }

  return { 
    newGrid, 
    scoreDelta: totalScore || 0, 
    penalty: penalty || 0, 
    laneScores, 
    replacedCells, 
    attachmentOffset,
    shiftedLanes 
  };
}

export function createEmptyGrid(size: number): RPS[][] {
  return Array.from({ length: size }, () => Array.from({ length: size }, () => RPS.BLANK));
}

File: godot/Scripts/Domain/Logic.cs
using System;
using System.Collections.Generic;

namespace Roshambo.Domain;

public static class Logic
{
	private static readonly Dictionary<Rps, Rps?> WinMap = new()
	{
		{ Rps.Rock, Rps.Scissors },
		{ Rps.Scissors, Rps.Paper },
		{ Rps.Paper, Rps.Rock },
		{ Rps.Blank, null },
		{ Rps.Tricolor, null }
	};

	private static readonly Dictionary<Rps, int> ScoreWeights = new()
	{
		{ Rps.Rock, 4 },
		{ Rps.Scissors, 3 },
		{ Rps.Paper, 1 },
		{ Rps.Blank, 0 },
		{ Rps.Tricolor, 2 }
	};

	private static Rps SymbolThatBeats(Rps target)
	{
		return target switch
		{
			Rps.Rock => Rps.Paper,
			Rps.Paper => Rps.Scissors,
			Rps.Scissors => Rps.Rock,
			_ => Rps.Rock
		};
	}

	private static Rps ResolveTricolor(Rps symbol, Rps opponent, bool wantsAdvantage)
	{
		if (symbol != Rps.Tricolor) return symbol;
		if (opponent == Rps.Blank) return Rps.Tricolor;
		if (!wantsAdvantage) return Rps.Rock;
		return SymbolThatBeats(opponent);
	}

	private static (Rps attacker, Rps defender) ResolvePair(Rps attacker, Rps defender)
	{
		if (attacker == Rps.Tricolor && defender == Rps.Tricolor)
		{
			return (Rps.Rock, Rps.Rock);
		}

		var resolvedAttacker = ResolveTricolor(attacker, defender, true);
		var resolvedDefender = ResolveTricolor(defender, resolvedAttacker, true);
		return (resolvedAttacker, resolvedDefender);
	}

	private static void ShiftLane(Rps[,] grid, int index, string type, int direction)
	{
		var size = grid.GetLength(0);

		if (type == "row")
		{
			if (direction == 1)
			{
				var last = grid[index, size - 1];
				for (var col = size - 1; col > 0; col--)
				{
					grid[index, col] = grid[index, col - 1];
				}
				grid[index, 0] = last;
			}
			else
			{
				var first = grid[index, 0];
				for (var col = 0; col < size - 1; col++)
				{
					grid[index, col] = grid[index, col + 1];
				}
				grid[index, size - 1] = first;
			}
		}
		else
		{
			if (direction == 1)
			{
				var last = grid[size - 1, index];
				for (var row = size - 1; row > 0; row--)
				{
					grid[row, index] = grid[row - 1, index];
				}
				grid[0, index] = last;
			}
			else
			{
				var first = grid[0, index];
				for (var row = 0; row < size - 1; row++)
				{
					grid[row, index] = grid[row + 1, index];
				}
				grid[size - 1, index] = first;
			}
		}
	}

	public static int ResolveAttachmentOffset(int matrixSize, int cardLength, float pointerRatio)
	{
		var minOffset = Math.Min(0, matrixSize - cardLength);
		var maxOffset = Math.Max(0, matrixSize - cardLength);
		if (minOffset == maxOffset) return minOffset;

		var placementCount = maxOffset - minOffset + 1;
		var clampedRatio = Math.Min(Math.Max(pointerRatio, 0f), 0.999999f);
		var placementIndex = Math.Min(placementCount - 1, (int)Math.Floor(clampedRatio * placementCount));
		return minOffset + placementIndex;
	}

	public static ClashResult ExecuteLaneClash(Rps[,] currentGrid, InsertEdge edge, Card card, int attachmentOffset)
	{
		var size = currentGrid.GetLength(0);
		var newGrid = new Rps[size, size];
		Array.Copy(currentGrid, newGrid, currentGrid.Length);

		var totalScore = 0;
		var penalty = 0;
		var laneScores = new int[card.Symbols.Length];
		var replacedCells = new List<ClashCell>();
		var failedCells = new List<ClashCell>();
		var tieCells = new List<ClashCell>();
		var shiftedLanes = new List<ShiftedLane>();

		for (var cardIndex = 0; cardIndex < card.Symbols.Length; cardIndex++)
		{
			var laneIndex = attachmentOffset + cardIndex;
			if (laneIndex < 0 || laneIndex >= size) continue;

			var attackerSymbol = card.Symbols[cardIndex];
			var r = 0;
			var c = 0;
			var dr = 0;
			var dc = 0;

			switch (edge)
			{
				case InsertEdge.Top: r = 0; c = laneIndex; dr = 1; dc = 0; break;
				case InsertEdge.Bottom: r = size - 1; c = laneIndex; dr = -1; dc = 0; break;
				case InsertEdge.Left: r = laneIndex; c = 0; dr = 0; dc = 1; break;
				case InsertEdge.Right: r = laneIndex; c = size - 1; dr = 0; dc = -1; break;
			}

			var defender = newGrid[r, c];
			var resolvedInitial = ResolvePair(attackerSymbol, defender);
			if (attackerSymbol == Rps.Tricolor)
			{
				attackerSymbol = resolvedInitial.attacker;
			}

			var attackerLoses = resolvedInitial.defender != Rps.Blank
				&& (resolvedInitial.attacker == Rps.Blank || (WinMap[resolvedInitial.defender] ?? Rps.Blank) == resolvedInitial.attacker);

			if (attackerLoses)
			{
				penalty += ScoreWeights.TryGetValue(resolvedInitial.attacker, out var attackerVal) ? attackerVal : 0;

				if (defender == Rps.Tricolor)
				{
					newGrid[r, c] = resolvedInitial.defender;
				}

				if (edge is InsertEdge.Left or InsertEdge.Right)
				{
					var shiftDir = edge == InsertEdge.Left ? -1 : 1;
					ShiftLane(newGrid, laneIndex, "row", shiftDir);
					shiftedLanes.Add(new ShiftedLane { Index = laneIndex, Type = "row", Direction = shiftDir });
				}
				else
				{
					var shiftDir = edge == InsertEdge.Top ? -1 : 1;
					ShiftLane(newGrid, laneIndex, "col", shiftDir);
					shiftedLanes.Add(new ShiftedLane { Index = laneIndex, Type = "col", Direction = shiftDir });
				}
			}
			else if (attackerSymbol != Rps.Blank)
			{
				for (var step = 0; step < size; step++)
				{
					var currentDefender = newGrid[r, c];
					var resolved = ResolvePair(attackerSymbol, currentDefender);
					var attackerWins = (WinMap[resolved.attacker] ?? Rps.Blank) == resolved.defender || resolved.defender == Rps.Blank;

					if (attackerWins)
					{
						totalScore += ScoreWeights.TryGetValue(resolved.defender, out var gain) ? gain : 0;
						laneScores[cardIndex] += ScoreWeights.TryGetValue(resolved.defender, out var laneGain) ? laneGain : 0;
						newGrid[r, c] = resolved.attacker;
						replacedCells.Add(new ClashCell { R = r, C = c });
						r += dr;
						c += dc;
						if (r < 0 || r > size - 1 || c < 0 || c > size - 1) break;
					}
					else
					{
						if ((WinMap[resolved.defender] ?? Rps.Blank) == resolved.attacker)
						{
							penalty += ScoreWeights.TryGetValue(resolved.attacker, out var p) ? p : 0;
							failedCells.Add(new ClashCell { R = r, C = c });
							if (currentDefender == Rps.Tricolor)
							{
								newGrid[r, c] = resolved.defender;
							}
						}
						else if (resolved.defender == resolved.attacker)
						{
							tieCells.Add(new ClashCell { R = r, C = c });
							if (currentDefender == Rps.Tricolor)
							{
								newGrid[r, c] = resolved.defender;
							}
						}

						break;
					}
				}
			}
		}

		return new ClashResult
		{
			NewGrid = newGrid,
			ScoreDelta = totalScore,
			Penalty = penalty,
			LaneScores = laneScores,
			ReplacedCells = replacedCells,
			FailedCells = failedCells,
			TieCells = tieCells,
			AttachmentOffset = attachmentOffset,
			ShiftedLanes = shiftedLanes
		};
	}

	public static Rps[,] CreateEmptyGrid(int size)
	{
		var grid = new Rps[size, size];
		for (var r = 0; r < size; r++)
		for (var c = 0; c < size; c++)
		{
			grid[r, c] = Rps.Blank;
		}
		return grid;
	}
}

